File: src/nfan/fan/sys/Enum.cs
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   29 Jan 07  Andy Frank  Creation
//

using Fanx.Fcode;

namespace Fan.Sys
{
  /// <summary>
  /// Enum base class.
  /// </summary>
  public abstract class Enum : FanObj
  {

  //////////////////////////////////////////////////////////////////////////
  // Construction
  //////////////////////////////////////////////////////////////////////////

    public static Enum make(Long ordinal, string name)
    {
      // should never be used
      throw new System.Exception();
    }

    public static void make_(Enum self, Long ordinal, string name)
    {
      if (ordinal == null || name == null) throw new NullErr().val;
      self.m_ordinal = ordinal;
      self.m_name    = name;
    }

    protected static Enum doFromStr(Type t, string name, Boolean check)
    {
      // the compiler marks the value fields with the Enum flag
      Slot slot = t.slot(name, false);
      if (slot != null && (slot.m_flags & FConst.Enum) != 0)
      {
        try
        {
          return (Enum)((Field)slot).get(null);
        }
        catch (System.Exception)
        {
        }
      }
      if (!check.booleanValue()) return null;
      throw ParseErr.make(t.qname(), name).val;
    }

  //////////////////////////////////////////////////////////////////////////
  // Identity
  //////////////////////////////////////////////////////////////////////////

    public override sealed Boolean _equals(object obj)
    {
      return this == obj ? Boolean.True : Boolean.False;
    }

    public override sealed Long compare(object obj)
    {
      return FanInt.compare(m_ordinal, ((Enum)obj).m_ordinal);
    }

    public override Type type()
    {
      return Sys.EnumType;
    }

    public override string toStr()
    {
      return m_name;
    }

    public Long ordinal()
    {
      return m_ordinal;
    }

    public string name()
    {
      return m_name;
    }

  //////////////////////////////////////////////////////////////////////////
  // Fields
  //////////////////////////////////////////////////////////////////////////

    private Long m_ordinal = null;
    private string m_name = null;

  }
}

File: src/jfan/fan/sys/Enum.java
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   16 Apr 06  Brian Frank  Creation
//
package fan.sys;

import fanx.fcode.*;

/**
 * Enum base class.
 */
public abstract class Enum
  extends FanObj
{

//////////////////////////////////////////////////////////////////////////
// Construction
//////////////////////////////////////////////////////////////////////////

  public static Enum make(Long ordinal, String name)
  {
    // should never be used
    throw new IllegalStateException();
  }

  public static void make$(Enum self, Long ordinal, String name)
  {
    if (ordinal == null || name == null) throw new NullErr().val;
    self.ordinal = ordinal;
    self.name    = name;
  }

  protected static Enum doFromStr(Type t, String name, boolean checked)
  {
    // the compiler marks the value fields with the Enum flag
    Slot slot = t.slot(name, false);
    if (slot != null && (slot.flags & FConst.Enum) != 0)
    {
      try
      {
        return (Enum)((Field)slot).get(null);
      }
      catch (Exception e)
      {
      }
    }
    if (!checked) return null;
    throw ParseErr.make(t.qname(), name).val;
  }

//////////////////////////////////////////////////////////////////////////
// Identity
//////////////////////////////////////////////////////////////////////////

  public final boolean _equals(Object obj)
  {
    return this == obj;
  }

  public final Long compare(Object obj)
  {
    return FanInt.compare(ordinal, ((Enum)obj).ordinal);
  }

  public Type type()
  {
    return Sys.EnumType;
  }

  public String toStr()
  {
    return name;
  }

  public final Long ordinal()
  {
    return ordinal;
  }

  public final String name()
  {
    return name;
  }

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  private Long ordinal;
  private String name;

}
